src/input-states.js

```diff
@@ -0,0 +1,289 @@
+    else if (!_.isArray(state)) return [];
+    else return port.listAttached().map(function(socketIndex){
+        return getPortStateIndex(component, vnid, portName, socketIndex);
+    });
+}
+
+/**
+ * Returns the state of this port for an identified socket. This can be used to
+ * look up the last state when checking to see if the state has changed.
+ * @param component a component facade
+ * @param vnid an identifier that distinguishes the set of states
+ * @param portName the name of the addressable/multi port for this component
+ * @param sourceNode the upstream component nodeId
+ * @param sourcePort the upstream component port name
+ */
+function getPortStateEntry(component, vnid, portName, sourceNode, sourcePort) {
+    if (!isMultiPort(component.inPorts[portName]))
+        throw Error("This port is not addressable/multi: " + portName);
+    var socketIndex = findSocketIndex(component, portName, sourceNode, sourcePort);
+    return getPortStateIndex(component, vnid, portName, socketIndex);
+}
+
+/**
+ * Returns the state of this port for the given socketIndex. This can be used to
+ * look up the last state when checking to see if the state has changed.
+ * Note that the socketIndex is not neccessarily the same as the array index
+ * @param component a component facade
+ * @param vnid an identifier that distinguishes the set of states
+ * @param portName the name of the addressable/multi port for this component
+ * @param socketIndex the socketIndex (or socketId) provided to a port data event
+ */
+function getPortStateIndex(component, vnid, portName, socketIndex) {
+    if (!isMultiPort(component.inPorts[portName]))
+        throw Error("This port is not addressable/multi: " + portName);
+    component.vnis = component.vnis || {};
+    component.vnis[vnid] = component.vnis[vnid] || {};
+    component.vnis[vnid][portName] = component.vnis[vnid][portName] || [];
+    if (component.vnis[vnid][portName].length)
+        return component.vnis[vnid][portName][socketIndex];
+    else if (vnid) return getPortStateIndex(component, '', portName, socketIndex);
+    else return undefined;
+}
+
+/**
+ * Changes the state of each given port. If a port name is missing it is unmodified.
+ * This can be run in a test environment to reset the state of the ports.
+ * @param component a component facade
+ * @param vnid an identifier that distinguishes the set of states
+ * @param portStates hash of portName to state
+ */
+function setAllPortStates(component, vnid, portStates) {
+    _.each(portStates, function(state, portName) {
+        setPortState(component, vnid, portName, state);
+    });
+    return component;
+}
+
+/**
+ * Changes the state of a port. When a new state is received on a port this
+ * should be called.
+ * @param component a component facade
+ * @param vnid an identifier that distinguishes the set of states
+ * @param portName name of the port on this component
+ * @param state the new state to assign this port to
```

Comment:
@param state the new state (or array of states if the port is multi/addressable) to assign to this port